cmd/kubeadm/app/util/apiclient/wait.go

```diff
@@ -114,33 +121,39 @@ func getControlPlaneComponents(cfg *kubeadmapi.ClusterConfiguration) []controlPl
 		portScheduler = fmt.Sprintf("%d", constants.KubeSchedulerPort)
 	}
 
-	addressAPIServer, idx := kubeadmapi.GetArgValue(cfg.APIServer.ExtraArgs, addressArg, -1)
+	addressAPIServer, idx := kubeadmapi.GetArgValue(cfg.APIServer.ExtraArgs, advertiseAddressArg, -1)
 	if idx == -1 {
-		addressAPIServer = defaultAddress
+		addressAPIServer = defaultAddressAPIServer
 	}
-	addressKCM, idx := kubeadmapi.GetArgValue(cfg.ControllerManager.ExtraArgs, addressArg, -1)
+	addressKCM, idx := kubeadmapi.GetArgValue(cfg.ControllerManager.ExtraArgs, bindAddressArg, -1)
 	if idx == -1 {
-		addressKCM = defaultAddress
+		addressKCM = defaultAddressKCM
 	}
-	addressScheduler, idx := kubeadmapi.GetArgValue(cfg.Scheduler.ExtraArgs, addressArg, -1)
+	addressScheduler, idx := kubeadmapi.GetArgValue(cfg.Scheduler.ExtraArgs, bindAddressArg, -1)
 	if idx == -1 {
-		addressScheduler = defaultAddress
+		addressScheduler = defaultAddressScheduler
 	}
 
-	urlFormat := "https://%s:%s/%s"
+	getURL := func(address, port, endpoint string) string {
+		return fmt.Sprintf(
+			"https://%s/%s",
+			net.JoinHostPort(address, port),
+			endpoint,
+		)
+	}
 	return []controlPlaneComponent{
-		{name: "kube-apiserver", url: fmt.Sprintf(urlFormat, addressAPIServer, portAPIServer, endpointLivez)},
-		{name: "kube-controller-manager", url: fmt.Sprintf(urlFormat, addressKCM, portKCM, endpointHealthz)},
-		{name: "kube-scheduler", url: fmt.Sprintf(urlFormat, addressScheduler, portScheduler, endpointLivez)},
+		{name: "kube-apiserver", url: getURL(addressAPIServer, portAPIServer, endpointLivez)},
+		{name: "kube-controller-manager", url: getURL(addressKCM, portKCM, endpointHealthz)},
+		{name: "kube-scheduler", url: getURL(addressScheduler, portScheduler, endpointLivez)},
```

Comment:
According to the plan, we will promote it into beta and enabled it by default in the next release. For the join phase, if one of the controller and scheduler use 127.0.0.1 as bind address, the check will always fail. so we need to know which phase we are in, and return the collections based on the phase and bind address.

Reply:
> According to the plan, we will promote it into beta and enabled it by default in the next release.

maybe for 1.33.

> For the join phase, if one of the controller and scheduler use 127.0.0.1 as bind address, the check will always fail. 

can you explain further?
the controller and scheduler always use 127.0.0.1 by default on all control plane nodes.
this e2e already tests a joining control plane node while the FG is enabled.
https://testgrid.k8s.io/sig-cluster-lifecycle-kubeadm#kubeadm-kinder-wait-for-all-control-plane-components-latest

this reminds me though of one related problem with this feature.
sometimes users use kubeadm --patches to patch the static pods and provide custom IPs for control plane components.
for that using the ClusterConfiguration as the source of overrides won't work.


patches are applied directly over the files on disk and we don't keep them in memory.
i think the right approach is to read the manifest file for a CP component on disk and parse the flag value.
we already have the tools for that.

Reply:
Sorry, it's my mistaken. I didn't notice that the phase would be skipped if joining a worker node.

```
	// Skip if this is not a control plane
	if data.Cfg().ControlPlane == nil {
		return nil
	}
```